src/util/future.rs

```diff
@@ -0,0 +1,29 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::boxed;
+use futures::sync::oneshot;
+
+/// Generated a paired future and callback so that when callback is being called, its result
+/// is automatically passed as a future result.
+pub fn gen_paired_future_callback<T>() -> (Box<boxed::FnBox(T) + Send>, oneshot::Receiver<T>)
+where
+    T: Send + 'static,
+{
+    let (tx, future) = oneshot::channel::<T>();
+    let callback = box move |result| {
+        let _ = tx.send(result);
```

Comment:
Just unwrap.

Reply:
Cannot unwrap here. `Result<_, E>`'s unwrap requires `E: Debug` so that it can be printed out when unwrap fails, which is not satisfied. Note: the signature is `fn send(&self, t: T) -> Result<(), T>`.

Reply:
In addition, we want to ignore errors here. Unwrap will produce a panic if it is an error.

Reply:
I prefer to fail fast, so let's panic.

Reply:
I suppose we may encounter send failures when shutting down, for example, there are remaining messages but the user is exiting the server. In such case there might be send failures, which we would like to ignore but not panic.

Reply:
Let's log the error.